oxide-vpc/src/engine/icmpv6.rs

```diff
@@ -155,24 +159,74 @@ fn add_icmpv6_rules(
     // And then pass through unchanged any ICMPv6 Echo Request / Reply messages.
     // These will not be matched by the ping rules above, meaning they're
     // destined for somewhere else. We can add more exceptions to the Deny above
-    // as needed. Note that we don't match on the IP addresses, expecting those
-    // to be handled at the router layer.
+    // as needed.
     let priority = u16::try_from(n_actions + 1).unwrap();
+
+    // Outbound echo messages.
     let mut rule = Rule::new(
         priority,
-        Action::Static(Arc::new(Identity::new("allow-icmpv6-echo"))),
+        Action::Static(Arc::new(Identity::new("allow-outbound-icmpv6-echo"))),
     );
+    // IPv6 only.
+    rule.add_predicate(Predicate::InnerEtherType(vec![EtherTypeMatch::Exact(
+        ETHER_TYPE_IPV6,
+    )]));
+    // From client MAC.
+    rule.add_predicate(Predicate::InnerEtherSrc(vec![EtherAddrMatch::Exact(
+        cfg.private_mac,
+    )]));
+    // To OPTE MAC.
+    rule.add_predicate(Predicate::InnerEtherDst(vec![EtherAddrMatch::Exact(
+        cfg.gateway_mac,
+    )]));
+    // From client IPs.
+    rule.add_predicate(Predicate::InnerSrcIp6(vec![
+        Ipv6AddrMatch::Exact(src_ips[0]),
+        Ipv6AddrMatch::Exact(src_ips[1]),
+    ]));
+    // ICMPv6
     rule.add_predicate(Predicate::InnerIpProto(vec![IpProtoMatch::Exact(
         Protocol::ICMPv6,
     )]));
+    // Supported message types.
     rule.add_data_predicate(DataPredicate::Icmpv6MsgType(MessageType::from(
         Icmpv6Message::EchoRequest,
     )));
     rule.add_data_predicate(DataPredicate::Icmpv6MsgType(MessageType::from(
         Icmpv6Message::EchoReply,
     )));
-    icmp.add_rule(Direction::In, rule.clone().finalize());
     icmp.add_rule(Direction::Out, rule.finalize());
 
+    // Inbound echo mesages.
+    let mut rule = Rule::new(
+        priority,
+        Action::Static(Arc::new(Identity::new("allow-inbound-icmpv6-echo"))),
+    );
+    // IPv6 only.
+    rule.add_predicate(Predicate::InnerEtherType(vec![EtherTypeMatch::Exact(
+        ETHER_TYPE_IPV6,
+    )]));
+    // To client MAC. Note we don't have a predicate on the source MAC.
+    rule.add_predicate(Predicate::InnerEtherDst(vec![EtherAddrMatch::Exact(
+        cfg.private_mac,
+    )]));
+    // To client IPs.
+    rule.add_predicate(Predicate::InnerDstIp6(vec![
+        Ipv6AddrMatch::Exact(src_ips[0]),
+        Ipv6AddrMatch::Exact(src_ips[1]),
```

Comment:
Same here, I don't see how a link-local address could reach from guest-to-guest or external-to-guest.